### PR TITLE
Reworks the lwap to make it better.

### PR DESCRIPTION
**Pull request by Qwertytoforty**
## What Does This PR Do

Lwap now crafts from a base laser vs the accelerator laser cannon. Same material cost, half the crafting time, crafting a gun twice is silly.

The LWAP now has a super advanced NT targeting scope in it. Capable of highlighting creatures in pitch black darkness, not even walls can stop the targeting. It takes 3 seconds to activate the scope's targeting.

However the scope is not perfect, and highlights mobs as dots. No diffrence between friend or foe. Could be a xenomorph, could be a vampire, could be a fellow officer. Communication is key.

LWAP projectile is now 25% faster unscoped, 2x scoped. This should make the longer shots possible. I might keep this at 2x for scope, will see how people feel. Otherwise without the speed boost, hitting a long shot on a mobile target is near impossible.

The LWAP, while in scoped mode and fully booted up, can also pierce ONE object / turf / wall. This makes it better for ambushes, be it in maints or shooting into head offices, but requires careful positioning. If it pierces something, it will not knock down.

Reworks piercing projectiles slightly so you can make them pierce a certain ammount of targets, rather than infinite. Only the lwap at the moment has a limit, of one when scoped in.

Lwap de-scopes when user moves.

## Why It's Good For The Game

The lwap is interesting. The scope is strong, but, it has a 5 second delay between shots. 90% of the time, the accelerator will be better for the same goal, and do much more damage. As  well, with long shots knocking down, it makes it near impossible to land 2 shots, as you have to click directly on the mob in order to hit them when crawling, and they will be moving, so you can't lead a shot.

This gives the weapon more of a niche as an dark cover weapon, where security can provide covering fire for hard to see targets, as the dots do not care about light, as well as an ambush weapons, using the dots / pierce to hit a vampire through maints when you get other officers to scare them towards you.

It also is sure to not outclass the accelerator or xray in their niche, the accelerator cannon still has much better DPS, and the xray pierce and dps is still better than the cannon, so for sustained fire xray will be better, or for blobs, but the lwap will be better for ambushes at long range.

For concern of people using it as thermals, it's shitty thermals, it needs you to stand still, and at that tech level you can probably get real thermal implants from rnd anyway.

## Images of changes

Will put some more in soon.

https://user-images.githubusercontent.com/52090703/232338223-3a50b4f1-b6b1-4b2b-80cb-51089ac3c232.mp4

##TODO:

- [x] Disable the scope when the user moves.
- [x] figure out how to center the targeting proc from the users view, vs having to do a double range from the user.

## Testing

Lots of sniper stuff, ensuring the pierce projectiles still work as normal

## Changelog
:cl:
tweak: LWAP now has a smart scope, which acts as a cheap thermal system when scoped in.
tweak: LWAP now when scoped in for a period of time can pierce one wall / mob / object.
tweak: LWAP now can be crafted from a base laser, rather than accelerator.
tweak: LWAP projectile speed is 25% faster and twice as fast when scoped in.
tweak: LWAP de-scopes when moving.
/:cl:

### CODE REVIEW

File: code/modules/projectiles/guns/energy/laser.dm [hunk 1]
```diff
@@ -113,43 +113,71 @@
...
+	if(ishuman(user))
+		var/mob/living/carbon/human/H = user
+		if(scope_active)
+			select_fire(H)
+			H.remove_status_effect(STATUS_EFFECT_LWAPSCOPE)
+			scope_active = FALSE
+		else if(do_after(user, 3 SECONDS, target = src))
+			if(zoomed) //We check after.
+				scope_active = TRUE
+				to_chat(user, "<b><span class='robot'>SMRT-SCOPE Online.</span></b>")
```
Comment (Henri215) [inline code/modules/projectiles/guns/energy/laser.dm:143]: Just to be sure, its supposed to be SMRT or SMART?

Comment (Qwertytoforty) [reply code/modules/projectiles/guns/energy/laser.dm:143]: whatever I decide to name it, SMRT seemed funny, but also have been asking around for ironic / good names for a "smart" scope

Comment (Cocacolagua) [reply code/modules/projectiles/guns/energy/laser.dm:143]: QCK-SCOPE

File: code/__DEFINES/status_effects.dm [hunk 2]
```diff
@@ -159,3 +159,5 @@
 
 /// gives you fluff messages for cough, sneeze, headache, etc but without an actual virus
 #define STATUS_EFFECT_FAKE_VIRUS /datum/status_effect/fake_virus
+
+#define STATUS_EFFECT_LWAPSCOPE /datum/status_effect/lwap_scope //This status effect lets the user see the lwap dots.
```
Comment (Sirryan2002) [inline code/__DEFINES/status_effects.dm:163]:
```suggestion
/// This status effect lets the user see the lwap dots.
#define STATUS_EFFECT_LWAPSCOPE /datum/status_effect/lwap_scope 
```

File: code/datums/status_effects/neutral.dm [hunk 3]
```diff
@@ -143,3 +143,81 @@
...
 	alert_type = null
+
+#define LWAP_LOCK_CAP 10
+
+/datum/status_effect/lwap_scope
+	id = "lwap_scope"
+	alert_type = null
+	duration = -1
+	tick_interval = 4
+	var/locks = 0
```
Comment (Sirryan2002) [inline code/datums/status_effects/neutral.dm:154]: can you document this var?

File: code/datums/status_effects/neutral.dm [hunk 4]
```diff
@@ -143,3 +143,81 @@
...
+	if(scope_turf)
+		for(var/mob/living/L in range(10, scope_turf))
+			if(locks >= LWAP_LOCK_CAP)
+				return
+			if(L == owner)
+				continue
+			if(L.stat == DEAD)
+				continue
+			if(isslime(L) || ismonkeybasic(L)) //xenobio moment
+				continue
```
Comment (Sirryan2002) [inline code/datums/status_effects/neutral.dm:181]:
```suggestion
			if(locks >= LWAP_LOCK_CAP)
				return
			if(L == owner || L.stat == DEAD || isslime(L) || ismonkeybasic(L)) //xenobio moment
				continue
```

File: code/datums/status_effects/neutral.dm [hunk 5]
```diff
@@ -143,3 +143,81 @@
...
+			new /obj/effect/temp_visual/lwap_ping(owner.loc, owner, L)
+			locks++
+			continue
```
Comment (Sirryan2002) [inline code/datums/status_effects/neutral.dm:184]:
```suggestion
```
dangling continue, loop will autocontinue once it reaches the end of the code block

File: code/modules/projectiles/guns/energy/laser.dm [hunk 6]
```diff
@@ -113,43 +113,94 @@
...
 	slot_flags = SLOT_BACK
 	can_holster = FALSE
 	weapon_weight = WEAPON_HEAVY
 	origin_tech = "combat=6;magnets=6;powerstorage=4"
-	ammo_type = list(/obj/item/ammo_casing/energy/laser/sniper)
+	ammo_type = list(/obj/item/ammo_casing/energy/laser/sniper, /obj/item/ammo_casing/energy/laser/sniper/pierce)
 	zoomable = TRUE
 	zoom_amt = 7
 	shaded_charge = TRUE
+	var/scope_active = FALSE //To ensure we don't activate the piercing ammo out of scope.
```
Comment (Sirryan2002) [inline code/modules/projectiles/guns/energy/laser.dm:130]: autodoc comment for this?

File: code/modules/projectiles/guns/energy/laser.dm [hunk 7]
```diff
@@ -113,43 +113,94 @@
...
+	START_PROCESSING(SSobj, src)
+
+/obj/item/gun/energy/lwap/Destroy()
+	STOP_PROCESSING(SSobj, src)
+	return ..()
+
+
+/obj/item/gun/energy/lwap/zoom(mob/living/user, forced_zoom)
+	. = ..()
+	if(ishuman(user))
```
Comment (Sirryan2002) [inline code/modules/projectiles/guns/energy/laser.dm:144]: make this a guard clause please

File: code/modules/projectiles/guns/energy/laser.dm [hunk 8]
```diff
@@ -113,43 +113,94 @@
...
+		else if(do_after(user, 3 SECONDS, target = src) && zoomed)
+			if(zoomed && !scope_active) //We check after to be sure.
+				scope_active = TRUE
+				to_chat(user, "<b><span class='robot'>SMRT-SCOPE Online.</span></b>")
+				select_fire(H)
+				H.apply_status_effect(STATUS_EFFECT_LWAPSCOPE, stored_dir)
+
+/obj/item/gun/energy/lwap/process()
+	. = ..()
+	if(isliving(loc))
```
Comment (Sirryan2002) [inline code/modules/projectiles/guns/energy/laser.dm:160]: make this a guard clause please

File: code/modules/projectiles/projectile_base.dm [hunk 9]
```diff
@@ -54,7 +54,7 @@
 	var/drowsy = 0
 	var/stamina = 0
 	var/jitter = 0
-	var/forcedodge = 0 //to pass through everything
+	var/forcedodge = 0 //Number of times an object can pass through an object. -1 is infinite
```
Comment (Sirryan2002) [inline code/modules/projectiles/projectile_base.dm:57]:
```suggestion
	/// Number of times an object can pass through an object. -1 is infinite
	var/forcedodge = 0 
```

File: code/datums/status_effects/neutral.dm (hunk 3 above)
Comment (Qwertytoforty) [reply code/datums/status_effects/neutral.dm:155]: Done!

File: code/modules/projectiles/guns/energy/laser.dm (hunk 7 above)
Comment (Qwertytoforty) [reply code/modules/projectiles/guns/energy/laser.dm]: Done!

File: code/modules/projectiles/guns/energy/laser.dm (hunk 8 above)
Comment (Qwertytoforty) [reply code/modules/projectiles/guns/energy/laser.dm]: Done!

**Review by Sirryan2002 — APPROVED** [review]

File: code/modules/projectiles/guns/energy/laser.dm (hunk 1 above)
Comment (Cocacolagua) [reply code/modules/projectiles/guns/energy/laser.dm]: actually nevermind SCOPE-CREEP is infinitely funnier and fits the ability more

File: code/modules/projectiles/guns/energy/laser.dm [hunk 10]
```diff
@@ -113,43 +113,97 @@
...
+			to_chat(user, "<b><span class='robot'>SCOPE_CREEPER_[rand(1, 9999)] Online.</span></b>")
+			select_fire(H)
+			H.apply_status_effect(STATUS_EFFECT_LWAPSCOPE, stored_dir)
+
+/obj/item/gun/energy/lwap/process()
+	. = ..()
+	if(!isliving(loc))
+		return
+	var/mob/living/M = loc
+	if(world.time - M.last_movement <= 21 && zoomed) //Not perfect, will cause moving just before scoping to not be happy, open for a better solution.
```
Comment (farie82) [inline code/modules/projectiles/guns/energy/laser.dm:165]: Magic number here `21`

Comment (Qwertytoforty) [reply code/modules/projectiles/guns/energy/laser.dm:165]: do we have a define for a process cycle, or should I just do 2.1 seconds.

File: code/modules/projectiles/guns/energy/laser.dm [hunk 11]
```diff
@@ -113,43 +113,97 @@
...
+	if(!ishuman(user))
+		return
+	var/mob/living/carbon/human/H = user
+	stored_dir = H.dir
+	if(scope_active && !zoomed)
+		select_fire(H)
+		H.remove_status_effect(STATUS_EFFECT_LWAPSCOPE)
+		scope_active = FALSE
+	else if(do_after(user, 3 SECONDS, target = src) && zoomed)
+		if(zoomed && !scope_active) //We check after to be sure.
```
[the remark this answers is not in the document]
Comment (Qwertytoforty) [reply code/modules/projectiles/guns/energy/laser.dm:154]: done

File: code/modules/projectiles/guns/energy/laser.dm [hunk 12]
```diff
@@ -113,43 +113,97 @@
...
+/obj/item/gun/energy/lwap/zoom(mob/living/user, forced_zoom)
+	. = ..()
+	if(!ishuman(user))
+		return
+	var/mob/living/carbon/human/H = user
+	stored_dir = H.dir
+	if(scope_active && !zoomed)
+		select_fire(H)
+		H.remove_status_effect(STATUS_EFFECT_LWAPSCOPE)
+		scope_active = FALSE
```
[the remark this answers is not in the document]
Comment (Qwertytoforty) [reply code/modules/projectiles/guns/energy/laser.dm:152]: done

File: code/modules/projectiles/guns/energy/laser.dm (hunk 10 above)
Comment (Qwertytoforty) [reply code/modules/projectiles/guns/energy/laser.dm:165]: Changed to 2.1 seconds + a comment

Comment (farie82) [reply code/modules/projectiles/guns/energy/laser.dm]: 2.1 SECONDS sounds a bit arbitrary, giving it a define/variable which you can name might help explain why you picked 2.1 SECONDS

**Review by farie82 — APPROVED** [review]

**Pull request merged**